// cotui/.componentTemplate/HelloJavascript/index.js
import $ from 'jquery';
import bootstrap from 'bootstrap';
import './resources/style.scss';
import * as COMMON from 'common';

class ___COMPONENTNAME____{
    constructor({target, title}={}, initDOM){
        this.name = 'cotui-___componentname____'
        this.target = target
        this.els = document.querySelectorAll(target)
        this.component_id = `${COMMON.utils.generateID()}`
        
        this.updateComponent = this.updateComponent.bind(this)
        
        this.title = title

        if(initDOM){
            this.els.forEach(($el)=>{
                const component = $el.querySelector('[is]')||$el
                if(component == null){
                    console.warn('Please set is attribute <div is="cotui-___componentname____" />')
                }
                if(component.getAttribute('is')==this.name){
                    const sections = component.querySelectorAll('.___componentname____')
                    sections.forEach(($section)=>{
                        const $title = $section.querySelector('.___componentname____--title');

                        this.title = $title.innerHTML;
                    })
                }
            })
        }


        this.init();
    }


    init(){
        this.els.forEach(($el)=>{
            //Mount component
            $el.innerHTML = this.___COMPONENTNAME____()
        })
    }

    updateComponent($el){
        $el.innerHTML = this.___COMPONENTNAME____()
    }


    ___COMPONENTNAME____(){
        return `
            <div cotui id="js-___componentname______${this.component_id}" >
                <div class="jumbotron">
                    <h1 class="display-4">${this.title} Component</h1>
                    <p>This is a sample</p>
                    <p class="lead">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus urna est, condimentum eget blandit ac, ultrices sit amet orci. Quisque eleifend mi non nisl ornare accumsan. Curabitur mi tellus, pellentesque ac eros in, placerat tincidunt lorem. Aliquam eget leo a ligula porta ornare. Aenean at nunc ac dolor aliquam sodales. Ut ante ex, congue finibus orci eu, rhoncus sagittis erat. Curabitur quis sem aliquet, commodo metus hendrerit, maximus lorem. Nam sit amet lacinia arcu. In pellentesque dictum lorem. Nam posuere lacus quis dignissim sagittis. Vivamus consequat luctus nisl sed finibus. Mauris non lacinia enim. Mauris luctus nisi vitae eros pharetra semper. Suspendisse pharetra vehicula ultricies. Phasellus at neque nisi. Fusce id nulla sit amet lacus consectetur rutrum ac quis felis.</p>
                </div>
            </div>
        `
    }




}

export default ___COMPONENTNAME____
