// cotui/.storybook/config.js
import { configure,addDecorator } from '@storybook/vue';
import { withOptions } from '@storybook/addon-options';


import Vue from 'vue';

/*
import Vue from 'vue';


const requireComponent = require.context(
  // The relative path of the components folder
  './components',
  // Whether or not to look in subfolders
  false,
  // The regular expression used to match base component filenames
  /Base[A-Z]\w+\.(vue|js)$/
)

requireComponent.keys().forEach(fileName => {
  // Get component config
  const componentConfig = requireComponent(fileName)

  // Get PascalCase name of component
  const componentName = upperFirst(
    camelCase(
      // Strip the leading `./` and extension from the filename
      fileName.replace(/^\.\/(.*)\.\w+$/, '$1')
    )
  )

  // Register component globally
  Vue.component(
    componentName,
    // Look for the component options on `.default`, which will
    // exist if the component was exported with `export default`,
    // otherwise fall back to module's root.
    componentConfig.default || componentConfig
  )
})
*/

// Option defaults:

addDecorator(
  withOptions({
    /**
     * name to display in the top left corner
     * @type {String} 
     */
    name: 'City of Toronto ',
    /**
     * URL for name in top left corner to link to
     * @type {String} 
     */
    url: '#',
    /**
     * show story component as full screen
     * @type {Boolean} 
     */
    goFullScreen: false,
    /**
     * display panel that shows a list of stories
     * @type {Boolean} 
     */
    showStoriesPanel: true,
    /**
     * display panel that shows addon configurations
     * @type {Boolean} 
     */
    showAddonPanel: true,
    /**
     * display floating search box to search through stories
     * @type {Boolean} 
     */
    showSearchBox: false,
    /**
     * show addon panel as a vertical panel on the right
     * @type {Boolean} 
     */
    addonPanelInRight: true,
    /**
     * sorts stories
     * @type {Boolean} 
     */
    sortStoriesByKind: false,
    /**
     * regex for finding the hierarchy separator
     * @example:
     *   null - turn off hierarchy
     *   /\// - split by `/`
     *   /\./ - split by `.`
     *   /\/|\./ - split by `/` or `.`
     * @type {Regex} 
     */
    hierarchySeparator: '\/',
    /**
     * regex for finding the hierarchy root separator
     * @example:
     *   null - turn off multiple hierarchy roots
     *   /\|/ - split by `|`
     * @type {Regex} 
     */
    hierarchyRootSeparator: null,
    /**
     * sidebar tree animations
     * @type {Boolean} 
     */
    sidebarAnimations: true,
    /**
     * id to select an addon panel
     * @type {String} 
     */
    selectedAddonPanel: undefined, // The order of addons in the "Addon panel" is the same as you import them in 'addons.js'. The first panel will be opened by default as you run Storybook
    /**
     * enable/disable shortcuts
     * @type {Boolean} 
     */
    enableShortcuts: true, // true by default
  })
);





const req = require.context('../src/stories', true, /\.js$/)

function loadStories() {
  req.keys().forEach((filename) => req(filename))
}

addDecorator(story=> {
  return {
    template:`
    <div class="container" style="padding: 1em">
      <story />
    </div>
    `
  }
});


configure(loadStories, module);
